### DISCUSSION TITLE
Insert large Json on database creation

### DISCUSSION THREAD

**Discussion by fbede** [simolus3/drift#3029] [Q&A] [answered]
In my app, I need to insert some default values into the db when the db is created. I am currently doing it by using onCreate on MigrationStrategy so that the defaults are always extracted from json and inserted whenever the db is created. Due to the size of the json that needs to be serialized I need to run it on a separate isolate because it's causing issues on the UI thread. I've tried using computeWithDatabase but I keep getting an error about illegal argument in isolate message. Please can anyone suggest alternative simple ways of doing things?

**Comment by simolus3** [comment]
If you post some more details about your approach and what the illegal argument actually was (recent Dart versions show which field was the problem), it would be easier to help.

An alternative way may be to simply [use an existing database](https://drift.simonbinder.eu/docs/examples/existing_databases/#using-an-existing-database): If the default values are the same for each user (and not loaded from a backend dynamically), you can create an asset containing an initialized database that already has the values in it. You can write a standalone Dart script (not part of your app) that uses your database class to insert the values once into a database file via `NativeDatabase`. Inside your app, you check if the database file exists before using it and extract it from that generated asset file if it doesn't.

But if you're using `NativeDatabase.createInBackground`, the actual SQL is being run in a background isolate already. So if you have e.g. a large SQL string containing data to insert, you can just use `customStatement` and not block the main isolate with that.

**Comment by fbede** [comment, marked as answer]
Thank you so much for your reply @simolus3 but I have already solved the issue. I later found out that I was trying to write to the instance of the database on the "main thread" instead of the one that was created by the new isolate. To help anyone who might make the same mistake I made, this is the code snippet that was causing the error:

```
await database.computeWithDatabase(
   
 computation: (database) async {
   
   final assetCurrencies = loadCurrenciesFromAsset(currenciesJson);
   
   final assetIcons = loadIconsFromAsset(iconsJson);
   
   await database.batch((batch) {
   
     batch
   
       ..insertAll(currencyTable, assetCurrencies)
   
       ..insertAll(assetIconTable, assetIcons);
   
   });
   
 },
   
 connect: LocalDB.test,
  
);
```

It was fixed by;
```
await database.computeWithDatabase(
   computation: (database) async {
   final assetCurrencies = loadCurrenciesFromAsset(currenciesJson);
   final assetIcons = loadIconsFromAsset(iconsJson);
   await database.batch((batch) {
     batch
          ..insertAll(database.currencyTable, assetCurrencies)
          ..insertAll(database.assetIconTable, assetIcons);
      });
    },
    connect: LocalDB.test,
  );
```

The difference is that in the code snippet where the code runs without issues, the table instances are gotten from the database provided in the callback.
